### ISSUE TITLE
Unnecessary memcopies emitted by autograd engine

### ISSUE THREAD

**Issue by ngimel** [pytorch/pytorch#5169]
- OS: Ubuntu 16.04
- PyTorch version: master
- How you installed PyTorch (conda, pip, source):source
- Python version:3.6
- CUDA/cuDNN version:9.0
- GPU models and configuration:
- GCC version (if compiling from source):

Using .view that actually changes dimensions in a model introduces 4(!) extra memcopies of the size of gradOutput in backward pass.  Minimum reproducer below, in the first loop where input is 3d and view is flattening it to 2D, there are 4 memcopies after softmax backward, in the second loop where view is a noop there is none. Running with emitted nvtx shows that the first memcopy comes from AsStridedBackward, 3 extra ones come from autogradCopySlices, it is possibly related to  #3970. This is a regression from 0.3 that did not have those memcopies regardless of the view. It is also likely that this issue affects memory usage, as we are not able to fit the models on master that used to fit on 0.3.  cc @colesbury 
```
import torch
import torch.nn as nn
import torch.nn.functional as F


class Model(nn.Module):
    def __init__(self):
        super().__init__()
        self.fc = torch.nn.Linear(256,256)
       
    def  forward(self, input, target):
        input = self.fc(input)
        input = input.view(-1, input.size(-1))
        loss = F.cross_entropy(input, target, size_average=False)
        return loss


input =torch.randn(20,400,256).cuda()
target = torch.ones(8000).cuda().long()
model = Model().cuda()    
for i in range(5):
    loss = model(torch.autograd.Variable(input, requires_grad = True), torch.autograd.Variable(target))
    loss.backward()

input =torch.randn(8000,256).cuda()
for i in range(5):
    loss = model(torch.autograd.Variable(input, requires_grad = True), torch.autograd.Variable(target))
    loss.backward()
```

**Comment by ngimel** [comment]
Bisection points to #3384 as the offending commit. cc @colesbury

**Comment by colesbury** [comment]
The issue is due to a combination of a 3d x 2d `matmul` (from the nn.Linear) with the behavior for 3384. 3384 added specific behavior for in-place operations on views.  The 3d x 2d case of matmul ends with a viewing operation:

https://github.com/pytorch/pytorch/blob/bf603299b6a61d05744dbee8a4ee642bbac2979f/aten/src/ATen/native/LinearAlgebra.cpp#L197

After the matmul, `linear` performs an in-place addition:

https://github.com/pytorch/pytorch/blob/5bcacb21d512c359a2a4e7f19d020b68b3103284/torch/nn/functional.py#L903-L906

The in-place operation on the view triggers the `CopySlices` backward.

There are two ways we can fix this:

1) Try to avoid copying in `CopySlices` and `AsStridedBackward` in favor of viewing operations when possible. This is a bit tricky, but doable at least for some common cases.
2) Specify an explicit derivative for matmul instead of differentiating the individual operations. This would avoid treating the output as a view of another Tensor.

(We can also do both)

**Comment by csarofeen** [comment]
Redefining nn/functional.py linear as (as you recommend in 2):
```.py
def linear(input, weight, bias=None):
   """                                                                                                                                                                                                    
   Applies a linear transformation to the incoming data: :math:`y = xA^T + b`.                                                                                                                            
                                                                                                                                                                                                           
   Shape:                                                                                                                                                                                                  
       - Input: :math:`(N, *, in\_features)` where `*` means any number of                                                                                                                                
         additional dimensions                                                                                                                                                                            
       - Weight: :math:`(out\_features, in\_features)`                                                                                                                                                    
       - Bias: :math:`(out\_features)`                                                                                                                                                                    
       - Output: :math:`(N, *, out\_features)`                                                                                                                                                            
   """

   if input.dim() > 2:
       input = input.contiguous()
       sizes = input.size()[:-1]
       input = input.view(-1, input.size(-1))
   if input.dim() == 2 and bias is not None:
       # fused op is marginally faster                                                                                                                                                                    
       output = torch.addmm(bias, input, weight.t())
       return output.view(*sizes, -1)

   output = input.matmul(weight.t())
   if bias is not None:
       output += bias
   return output
```
Was enough to avoid the mem copies in fairseq. Should we just route though to addmm for now until someone tackles 1?

**Comment by colesbury** [comment]
@csarofeen that might work. I think fairseq uses 3d input, so you may need some viewing before and after the addmm call.  That should be fine as long as the bias addition is part of the `addmm` call and not a subsequent in-place addition.

**Comment by ngimel** [comment]
That's exactly what we are doing (viewing before and after) and we've verified that it works for fairseq (no ooms, performance back at 0.3 levels). We'll submit a PR then.

**Comment by soumith** [comment]
https://github.com/pytorch/pytorch/pull/5279 is merged now.
